src/SubPoolFactory.sol

```diff
@@ -0,0 +1,196 @@
+
+        return address(subpool);
+    }
+
+    /// @notice Bill a fine to a subpool
+    /// @param pool   - The subpool to fine.
+    /// @param amt    - Amount of pool's collateralToken to fine.
+    /// @param cowAmt - Amount of COW to fine.
+    /// @param ethAmt - Amount of ETH to fine.
+    /// @param reason - Reason to fine.
+    function bill(address pool, uint256 amt, uint256 cowAmt, uint256 ethAmt, string calldata reason) external auth {
+        SubPool(pool).bill(amt, cowAmt, ethAmt, msg.sender);
+        emit SolverPoolBilled(pool, amt, cowAmt, ethAmt, reason);
+    }
+
+    /// @notice Announce the intent to exit solving batch auctions.
+    function announceExit() external {
+        address pool = msg.sender;
+        SubPoolData memory subpoolData = subPoolData[pool];
+        if (subpoolData.collateral == address(0)) revert SubPoolFactory__UnknownPool();
+        if (subpoolData.exitTimestamp != 0) revert SubPoolFactory__PoolAlreadyAnnouncedExit();
+        subPoolData[pool].exitTimestamp = uint88(block.timestamp + cfg.exitDelay);
+        emit AnnounceExit(pool);
+    }
+
+    /// @notice Exit the pool.
+    function exitPool() external {
+        address pool = msg.sender;
+        SubPoolData memory subpoolData = subPoolData[pool];
+        if (subpoolData.collateral == address(0)) revert SubPoolFactory__UnknownPool();
+        if (subpoolData.exitTimestamp == 0) revert SubPoolFactory__PoolHasNotAnnouncedExitYet();
+        if (subpoolData.exitTimestamp > block.timestamp) revert SubPoolFactory__ExitDelayNotElapsed();
+        if (subpoolData.hasExited) revert SubPoolFactory__PoolAlreadyExited();
+        subPoolData[pool].hasExited = true;
+        emit Exit(pool);
+    }
+
+    /// @notice Update solver's backend uri.
+    function updateBackendUri(string calldata uri) external {
+        address pool = msg.sender;
+        SubPoolData memory subpoolData = subPoolData[pool];
+        if (subpoolData.collateral == address(0)) revert SubPoolFactory__UnknownPool();
+        backendUri[pool] = uri;
+        emit UpdateBackendUri(pool, uri);
+    }
+
+    /// @notice Override the exit timestamp to allow for an earlier exit.
+    function fastTrackExit(address pool, uint88 newExitTimestamp) external auth {
+        SubPoolData memory subpoolData = subPoolData[pool];
+        if (subpoolData.collateral == address(0)) revert SubPoolFactory__UnknownPool();
+        if (subpoolData.exitTimestamp == 0) revert SubPoolFactory__PoolHasNotAnnouncedExitYet();
+        if (subpoolData.hasExited) revert SubPoolFactory__PoolAlreadyExited();
+        if (newExitTimestamp > subpoolData.exitTimestamp) revert SubPoolFactory__InvalidFastTrackExit();
+        subpoolData.exitTimestamp = newExitTimestamp;
+        subPoolData[pool] = subpoolData;
+    }
+
+    /// @notice Determine if the solver can submit solutions.
+    function canSolve(address solver) external view returns (bool) {
+        address pool = poolOf(solver);
```

Comment:
This means that each pool can only be associated with one solver address (and they have to seed the pool through that address which has to be an EOA since it's the one later sending the transactions).

Is it possible to store a list of submission account -> pool so that one pool has multiple submission accounts (since `canSolve` is called from off-chain gas for traversing over a list shouldn't be so bad)

Reply:
> will then subsequently require an additional SLOAD in the signedSettle execution path.

How so? I was under the assumption `canSolve` was called offchain and the only thing we check in `signedSettle` is that the signature was given by some immutable address (which is part of the wrapper initialisation) for msg.sender ?

We don't have to call `canSolve` in a transaction ever imo.

Reply:
Implemented this